src/services/exportInfoMap.ts

```diff
@@ -0,0 +1,411 @@
+        get(importingFile: Path, importedName: string, symbol: Symbol, moduleName: string, checker: TypeChecker): readonly SymbolExportInfo[] | undefined;
+        forEach(importingFile: Path, action: (info: readonly SymbolExportInfo[], name: string, isFromAmbientModule: boolean) => void): void;
+        releaseSymbols(): void;
+        isEmpty(): boolean;
+        /** @returns Whether the change resulted in the cache being cleared */
+        onFileChanged(oldSourceFile: SourceFile, newSourceFile: SourceFile, typeAcquisitionEnabled: boolean): boolean;
+    }
+
+    export interface CacheableExportInfoMapHost {
+        getCurrentProgram(): Program | undefined;
+        getPackageJsonAutoImportProvider(): Program | undefined;
+    }
+
+    export function createCacheableExportInfoMap(host: CacheableExportInfoMapHost): ExportInfoMap {
+        let exportInfoId = 1;
+        const exportInfo = createMultiMap<string, CachedSymbolExportInfo>();
+        const symbols = new Map<number, [symbol: Symbol, moduleSymbol: Symbol]>();
+        let usableByFileName: Path | undefined;
+        const cache: ExportInfoMap = {
+            isUsableByFile: importingFile => importingFile === usableByFileName,
+            isEmpty: () => !exportInfo.size,
+            clear: () => {
+                exportInfo.clear();
+                symbols.clear();
+                usableByFileName = undefined;
+            },
+            add: (importingFile, symbol, moduleSymbol, moduleFile, exportKind, isFromPackageJson, scriptTarget, checker) => {
+                if (importingFile !== usableByFileName) {
+                    cache.clear();
+                    usableByFileName = importingFile;
+                }
+                const isDefault = exportKind === ExportKind.Default;
+                const importedName = getNameForExportedSymbol(isDefault && getLocalSymbolForExportDefault(symbol) || symbol, scriptTarget);
+                const moduleName = stripQuotes(moduleSymbol.name);
+                const id = exportInfoId++;
+                const storedSymbol = symbol.flags & SymbolFlags.Transient ? undefined : symbol;
+                const storedModuleSymbol = moduleSymbol.flags & SymbolFlags.Transient ? undefined : moduleSymbol;
+                if (!storedSymbol || !storedModuleSymbol) symbols.set(id, [symbol, moduleSymbol]);
+
+                exportInfo.add(key(importedName, symbol, moduleName, checker), {
+                    id,
+                    symbolName: importedName,
+                    moduleName,
+                    moduleFile,
+                    moduleFileName: moduleFile?.fileName,
+                    exportKind,
+                    isTypeOnly: !(skipAlias(symbol, checker).flags & SymbolFlags.Value),
+                    isFromPackageJson,
+                    symbol: storedSymbol,
+                    moduleSymbol: storedModuleSymbol,
+                });

```

Comment:
Symbols stored in `symbols` if transient, `exportInfo` if not. I tried never storing _any_ symbols, but the performance hit was too great.